File: microsynth/microsynth/report/label_finder/label_finder.py
# Copyright (c) 2024, Microsynth, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
import re


def get_columns(filters):
    return [
        {"label": _("Label Barcode"), "fieldname": "label_id", "fieldtype": "Data", "width": 100 },
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 75 },
        {"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 75 },
        {"label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 175 },
        {"label": _("Contact"), "fieldname": "contact", "fieldtype": "Link", "options": "Contact", "width": 75 },
        {"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order", "width": 125 },
        {"label": _("Web Order ID"), "fieldname": "web_order_id", "fieldtype": "Data", "width": 95 },
        {"label": _("Item Code"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 75 },
        {"label": _("Registered"), "fieldname": "registered", "fieldtype": "Check", "width": 80 },
        {"label": _("Registered To"), "fieldname": "registered_to", "fieldtype": "Link", "options": "Contact", "width": 100 },
        {"label": _("Sequencing Label"), "fieldname": "name", "fieldtype": "Link", "options": "Sequencing Label", "width": 95 }
    ]


def get_data(filters):
    if not filters:
        return []
    conditions = ""
    
    if filters.get('contact'):
        conditions += f"AND `tabSequencing Label`.`contact` = '{filters.get('contact')}'"
    if filters.get('registered_to'):
        conditions += f"AND `tabSequencing Label`.`registered_to` = '{filters.get('registered_to')}'"
    if filters.get('customer'):
        conditions += f"AND `tabSequencing Label`.`customer` = '{filters.get('customer')}'"
    if filters.get('customer_name'):
        conditions += f"AND `tabSequencing Label`.`customer_name` LIKE '{filters.get('customer_name')}'"
    if filters.get('sales_order'):
        conditions += f"AND `tabSequencing Label`.`sales_order` = '{filters.get('sales_order')}'"
    if filters.get('web_order_id'):
        conditions += f"AND `tabSales Order`.`web_order_id` = '{filters.get('web_order_id')}'"
    if filters.get('from_barcode') and filters.get('to_barcode'):
        if filters.get('from_barcode').isnumeric() and filters.get('to_barcode').isnumeric():
            conditions += f"AND `tabSequencing Label`.`label_id` BETWEEN '{filters.get('from_barcode')}' AND '{filters.get('to_barcode')}'"
        else:
            from_prefix = ''.join([i for i in filters.get('from_barcode') if not i.isdigit()])
            to_prefix = ''.join([i for i in filters.get('to_barcode') if not i.isdigit()])
            if from_prefix != to_prefix:
                frappe.throw("From Barcode and To Barcode need to have the same Prefix.")
            from_barcode = re.sub("[^0-9]", "", filters.get('from_barcode'))
            to_barcode = re.sub("[^0-9]", "", filters.get('to_barcode'))
            if len(from_barcode) != len(to_barcode):
                frappe.throw("From Barcode and To Barcode need to have the same length.")
            barcode_list = ','.join(f'"{to_prefix}{i:0{len(to_barcode)}d}"' for i in range(int(from_barcode), int(to_barcode) + 1))
            conditions += f"AND `tabSequencing Label`.`label_id` IN ({barcode_list})"

    sql_query = f"""
        SELECT
            `tabSequencing Label`.`name`,
            `tabSequencing Label`.`status`,
            `tabSequencing Label`.`registered`,
            `tabSequencing Label`.`label_id`,
            `tabSequencing Label`.`item` AS `item_code`,
            `tabSequencing Label`.`customer`,
            `tabSequencing Label`.`customer_name`,
            `tabSequencing Label`.`sales_order`,
            `tabSales Order`.`web_order_id`,
            `tabSequencing Label`.`contact`,
            `tabSequencing Label`.`registered_to`
        FROM `tabSequencing Label`
        LEFT JOIN `tabSales Order` ON `tabSales Order`.`name` = `tabSequencing Label`.`sales_order`
        WHERE TRUE
            {conditions}
        ORDER BY `tabSequencing Label`.`label_id`;
        """
    data = frappe.db.sql(sql_query, as_dict=True)
    return data


def execute(filters):
    columns, data = get_columns(filters), get_data(filters)
    return columns, data

File: microsynth/microsynth/report/label_finder/label_finder.js
// Copyright (c) 2024, Microsynth, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Label Finder"] = {
    "filters": [
        {
            "fieldname": "contact",
            "label": __("Contact"),
            "fieldtype": "Link",
            "options": "Contact"
        },
        {
            "fieldname": "registered_to",
            "label": __("Registered To"),
            "fieldtype": "Link",
            "options": "Contact"
        },
        {
            "fieldname": "customer",
            "label": __("Customer ID"),
            "fieldtype": "Link",
            "options": "Customer"
        },
        {
            "fieldname": "customer_name",
            "label": __("Customer Name"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "from_barcode",
            "label": __("From Barcode"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "to_barcode",
            "label": __("To Barcode"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "sales_order",
            "label": __("Sales Order"),
            "fieldtype": "Link",
            "options": "Sales Order"
        },
        {
            "fieldname": "web_order_id",
            "label": __("Web Order ID"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "item_code",
            "label": __("Item Code"),
            "fieldtype": "Link",
            "options": "Item"
        },
        {
            "fieldname": "registered",
            "label": __("Registered"),
            "fieldtype": "Check"
        }
    ],
    "onload": (report) => {
        hide_chart_buttons();
        hide_column_filters()
        report.page.add_inner_button( __("Lock Labels"), function() {
            var labels_to_lock = [];
            // check that all labels are unused
            var all_labels_unused = true;
            for (var i = 0; i < frappe.query_report.data.length; i++) {
                if (frappe.query_report.data[i].status != "unused") {
                    frappe.msgprint("The Sequencing Label " + frappe.query_report.data[i].name + " with Barcode " + frappe.query_report.data[i].label_id + " has Status " + frappe.query_report.data[i].status + ". Unable to set status to 'locked'. No Label status was changed. Please contact IT App if you have a valid use case.");
                    all_labels_unused = false;
                    break;
                }
                labels_to_lock.push({
                    'label_id': frappe.query_report.data[i].label_id,
                    'item_code': frappe.query_report.data[i].item_code
                });
            }
            if (all_labels_unused) {
                // TODO: ask for a reason
                frappe.msgprint("This functionality is not yet implemented.");
                // trigger label locking
                // frappe.confirm('Are you sure you want to <b>lock</b> the ' + frappe.query_report.data.length + ' Sequencing Labels selected in the Label Finder?',
                //     () => {
                //         frappe.call({
                //             'method': "microsynth.microsynth.seqblatt.lock_labels",
                //             'args':{
                //                 'content': {'labels': labels_to_lock}
                //             },
                //             'freeze': true,
                //             'freeze_message': __("Locking Labels ..."),
                //             'callback': function(r)
                //             {
                //                 frappe.show_alert('Locked Labels');
                //                 frappe.click_button('Refresh');
                //             }
                //         });
                //     }, () => {
                //         frappe.show_alert('No changes made');
                // });
            }
        });
        report.page.add_inner_button( __("Set Labels unused"), function() {
            var labels_to_set_unused = [];
            // check that all labels are locked
            var all_labels_locked = true;
            for (var i = 0; i < frappe.query_report.data.length; i++) {
                if (frappe.query_report.data[i].status != "locked") {
                    frappe.msgprint("The Sequencing Label " + frappe.query_report.data[i].name + " with Barcode " + frappe.query_report.data[i].label_id + " has Status " + frappe.query_report.data[i].status + ". Unable to set status to 'unused'. No Label status was changed. Please contact IT App if you have a valid use case.");
                    all_labels_locked = false;
                    break;
                }
                labels_to_set_unused.push({
                    'label_id': frappe.query_report.data[i].label_id,
                    'item_code': frappe.query_report.data[i].item_code
                });
            }
            if (all_labels_locked) {
                // TODO: ask for a reason
                frappe.msgprint("This functionality is not yet implemented.");
                // set Labels to status unused
                // frappe.confirm('Are you sure you want to set the ' + frappe.query_report.data.length + ' Sequencing Labels selected in the Label Finder to status <b>unused</b>?',
                //     () => {
                //         frappe.call({
                //             'method': "microsynth.microsynth.seqblatt.set_unused",
                //             'args':{
                //                 'content': {'labels': labels_to_set_unused}
                //             },
                //             'freeze': true,
                //             'freeze_message': __("Setting Labels to unused ..."),
                //             'callback': function(r)
                //             {
                //                 frappe.show_alert('Set Labels to unused');
                //                 frappe.click_button('Refresh');
                //             }
                //         });
                //     }, () => {
                //         frappe.show_alert('No changes made');
                // });
            }
        });
    }
};

function hide_column_filters() {
    let container = document.getElementsByClassName("page-content");
    const hide_column_filter_style = document.createElement("style");
    hide_column_filter_style.innerHTML = `
        .dt-header .dt-row[data-is-filter] {
          display: none !important;
        }
    `
    for (let i = 0; i < container.length; i++) {
        container[i].appendChild(hide_column_filter_style);
    }
}
